### ISSUE TITLE
Failed to perform add on <Link: PirAnka - PirAnka>

### ISSUE THREAD

**Issue by acoul** [netjson/django-netjsongraph#58]
I am getting sporadic errors in the log. Looking at the [NetJSON](http://dev.ozo.com/awmn.json) input I can't locate an entry that justifies this error.

the error message is this:

> [ERROR 2017-09-22 09:20:14,358] module: contextmanagers, process: 27803, thread: -1213786368
Failed to perform add on <Link: PirAnka - PirAnka>
Traceback (most recent call last):
  File "/extra/inet/django-netjsongraph/src/django-netjsongraph/django_netjsongraph/contextmanagers.py", line 10, in log_failure
    yield
  File "/extra/inet/django-netjsongraph/src/django-netjsongraph/django_netjsongraph/base/topology.py", line 250, in update
    link.full_clean()
  File "/usr/lib/python3.4/site-packages/django/db/models/base.py", line 1248, in full_clean
    raise ValidationError(errors)
django.core.exceptions.ValidationError: {'__all__': ['source and target must not be the same']}

**Comment by nemesisdesign** [comment]
@acoul would the signature of the object that is failing help you locate the issue?

**Comment by nemesisdesign** [comment]
Let's clarify, what do you think the issue is?

1. the error message in the log is not good enough hence not helpful
2. the fact that a validation error is raised if a node has a link to itself doesn't work in your use case

**Comment by acoul** [comment]
looking @ the NetJSON file I can't locate a link with same target & source. I don;t understand what is triggering this error/behavior. I've made a smaller subset of [NetJSON file](http://spq.ozo.com/awmn-test.json) that causes the same issue.

**Comment by nemesisdesign** [comment]
@acoul I performed some tests and indeed there's something fishy going on.
Even trying to improve the error message won't help.
When this kind of weird bugs pop up, my usual resort is the brute force strategy: the NetJSON subset should be decreased iteratively until the smallest possible subset that triggers the error remains. Can you do this? If you can that would help a lot: I will be able to run very fast tests on the small subset in a semi-automated manner, if a bug is indeed found I could even use that sample in the automated test suite.

**Comment by acoul** [comment]
I created a very [small NetJSON subset](http://spq.ozo.com/awmn-test.json) that recreates this issue. The NetJSON is been generated by this simple set of ASs:

> 17101 19332
17101 19479
17101 20802
17101 21042
18569 18606
18569 19835
18569 21194
18570 19835
44038 9895
44038 19835
44038 22239

hope this helps

**Comment by nemesisdesign** [comment]
@acoul trying with that dataset I get a similar error:

```
Failed to perform add on <Link: SV1CIM-SOFIKO-SV1GSD - SV1CIM-SOFIKO-SV1GSD>: ValidationError
```

Is this the same error you are getting with the same dataset? If yes I can proceed with some more testing.

**Comment by acoul** [comment]
> [ERROR 2017-09-27 18:56:19,273] module: contextmanagers, process: 7641, thread: -1213450496
Failed to perform add on <Link: SV1CIM-SOFIKO-SV1GSD - SV1CIM-SOFIKO-SV1GSD>
Traceback (most recent call last):
  File "/extra/inet/django-netjsongraph/src/django-netjsongraph/django_netjsongraph/contextmanagers.py", line 10, in log_failure
    yield
  File "/extra/inet/django-netjsongraph/src/django-netjsongraph/django_netjsongraph/base/topology.py", line 250, in update
    link.full_clean()
  File "/usr/lib/python3.4/site-packages/django/db/models/base.py", line 1248, in full_clean
    raise ValidationError(errors)
django.core.exceptions.ValidationError: {'__all__': ['source and target must not be the same']}

yes ACK. this is the error I am getting too.

**Comment by nemesisdesign** [comment]
@acoul I prepared a patch in pull request #58 (branch `issues/58`).

Before merging it in master I would like us to test it a bit. Be sure to update your django-netjsongraph code with the code on that branch and apply the data migration with `./manage.py migrate`.

While working on this issue I realized a better long term solution is needed and opened issue #61.

**Comment by acoul** [comment]
I am still getting the same error. Here are the steps I followed:

> cd django-netjsongraph; git pull; git checkout issues/58; pip install ./ --user; cd ~web/nmsbgp; ./manage.py makemigrations; ./manage.py migrate

I just added the "./manage.py makemigrations" step as I was instructed by the following message:

>   Your models have changes that are not yet reflected in a migration, and so won't be applied.
  Run 'manage.py makemigrations' to make new migrations, and then re-run 'manage.py migrate' to apply them.

here is the error message:

> [ERROR 2017-10-03 13:05:45,093] module: contextmanagers, process: 14240, thread: -1214155008
Failed to perform add on <Link: SV1CIM-SOFIKO-SV1GSD - SV1CIM-SOFIKO-SV1GSD>
Traceback (most recent call last):
  File "/extra/inet/django-netjsongraph/src/django-netjsongraph/django_netjsongraph/contextmanagers.py", line 10, in log_failure
    yield
  File "/extra/inet/django-netjsongraph/src/django-netjsongraph/django_netjsongraph/base/topology.py", line 250, in update
    link.full_clean()
  File "/usr/lib/python3.4/site-packages/django/db/models/base.py", line 1248, in full_clean
    raise ValidationError(errors)
django.core.exceptions.ValidationError: {'__all__': ['source and target must not be the same']}

**Comment by acoul** [comment]
I decided to delete and recreate the database & the sub categories. The error persist with some additional nodes:

> [ERROR] Failed to perform add on <Link: PirAnka - PirAnka>
Traceback (most recent call last):
  File "/extra/inet/django-netjsongraph/src/django-netjsongraph/django_netjsongraph/contextmanagers.py", line 10, in log_failure
    yield
  File "/extra/inet/django-netjsongraph/src/django-netjsongraph/django_netjsongraph/base/topology.py", line 250, in update
    link.full_clean()
  File "/usr/lib/python3.4/site-packages/django/db/models/base.py", line 1248, in full_clean
    raise ValidationError(errors)
django.core.exceptions.ValidationError: {'__all__': ['source and target must not be the same']}
Failed to perform add on <Link: PirAnka - PirAnka>: ValidationError
See error log for more information

> [ERROR] Failed to perform add on <Link: Benz_Kalloni - Benz_Kalloni>
Traceback (most recent call last):
  File "/extra/inet/django-netjsongraph/src/django-netjsongraph/django_netjsongraph/contextmanagers.py", line 10, in log_failure
    yield
  File "/extra/inet/django-netjsongraph/src/django-netjsongraph/django_netjsongraph/base/topology.py", line 250, in update
    link.full_clean()
  File "/usr/lib/python3.4/site-packages/django/db/models/base.py", line 1248, in full_clean
    raise ValidationError(errors)
django.core.exceptions.ValidationError: {'__all__': ['source and target must not be the same']}
Failed to perform add on <Link: Benz_Kalloni - Benz_Kalloni>: ValidationError
See error log for more information

> [ERROR] Failed to perform add on <Link: SV1CIM-SOFIKO-SV1GSD - SV1CIM-SOFIKO-SV1GSD>
Traceback (most recent call last):
  File "/extra/inet/django-netjsongraph/src/django-netjsongraph/django_netjsongraph/contextmanagers.py", line 10, in log_failure
    yield
  File "/extra/inet/django-netjsongraph/src/django-netjsongraph/django_netjsongraph/base/topology.py", line 250, in update
    link.full_clean()
  File "/usr/lib/python3.4/site-packages/django/db/models/base.py", line 1248, in full_clean
    raise ValidationError(errors)
django.core.exceptions.ValidationError: {'__all__': ['source and target must not be the same']}
Failed to perform add on <Link: SV1CIM-SOFIKO-SV1GSD - SV1CIM-SOFIKO-SV1GSD>: ValidationError
See error log for more information

also the orphan node does appear again as orphan in the new graph as reported on https://github.com/netjson/django-netjsongraph/issues/59

**Comment by acoul** [comment]
in case this may help, it seems there is a relationship between the names of the Nodes that get the above errors. searching with grep on the json file I came up with the following:
    }, 
    {
      "source": "ANKA", 
      "cost": 1.0, 
      "target": "ANKA2", 
      "cost_text": "", 
      "properties": {}
    }, 
    {
      "source": "ANKA", 
      "cost": 1.0, 
      "target": "PirAnka", 
      "cost_text": "", 
      "properties": {}
    }, 
    {
      "source": "Kalloni", 
      "cost": 1.0, 
      "target": "Benz_Kalloni_2", 
      "cost_text": "", 
      "properties": {}
    }, 
    {
      "source": "Kalloni", 
      "cost": 1.0, 
      "target": "Benz_Kalloni", 
      "cost_text": "", 
      "properties": {}
    }, 
    {
      "source": "SV1CIM-SOFIKO-SV1GSD", 
      "cost": 1.0, 
      "target": "Sv1gyk-Sofiko", 
      "cost_text": "", 
      "properties": {}
    }, 
    {
      "source": "sv1gsd_2", 
      "cost": 1.0, 
      "target": "SV1CIM-SOFIKO-SV1GSD", 
      "cost_text": "", 
      "properties": {}
    }, 
    {
      "source": "sv1gsd", 
      "cost": 1.0, 
      "target": "SV1CIM-SOFIKO-SV1GSD", 
      "cost_text": "", 
      "properties": {}

**Comment by nemesisdesign** [comment]
@acoul yes it's a problem related to the NetJSON `id` and `local_address` fields.
The shortcut I initially used to implement this was a bad choice (issue #61).

The bug is not in the parser but in the methods that look for existing links and nodes.

I thought the patched solved the problem but evidently I did not have enough data to test, I will try another patch as soon as possible and let you know.

**Comment by acoul** [comment]
Today I synced to latest git tree. I had to migrate after that in order to update the database. I then deleted all nodes & links from the admin panel and then I run update_topology.

I got some of the errors mentioned in this ticket. I then merged the changes in the issue/58 branch to the master tree, installed the package with pip install ./ -U, deleted all nodes & links from the admin panel and then I run update_topology which produced no errors.
